src/passes/StringLowering.cpp

```diff
@@ -0,0 +1,188 @@
+        stringPtrs.push_back(stringPtr);
+      }
+    }
+
+    // Generate the indexes from the combined set of necessary strings, which we
+    // sort for determinism (alphabetically).
+    for (auto& string : stringSet) {
+      strings.push_back(string);
+    }
+    std::sort(strings.begin(), strings.end());
+    for (Index i = 0; i < strings.size(); i++) {
+      stringIndexes[strings[i]] = i;
+    }
+  }
+
+  // For each string index, the name of the global that replaces it.
+  std::vector<Name> globalNames;
+
+  Type nnstringref = Type(HeapType::string, NonNullable);
+
+  // Existing globals already in the form we emit can be reused. That is, if
+  // we see
+  //
+  //  (global $foo (ref null string) (string.const ..))
+  //
+  // then we can just use that as the global for that string. This avoids
+  // repeated executions of the pass adding more and more globals.
+  //
+  // Note that we don't note these in newNames: They are already in the right
+  // sorted position, before any uses, as we use the first of them for each
+  // string. Only actually new names need sorting.
+  //
+  // Any time we reuse a global, we must not modify its body (or else we'd
+  // replace the global that all others read from); we note them here and
+  // avoid them in replaceStrings later to avoid such trampling.
+  std::unordered_set<Expression**> stringPtrsToPreserve;
+
+  void addGlobals(Module* module) {
+    // Note all the new names we create for the sorting later.
+    std::unordered_set<Name> newNames;
+
+    // Each string will get a global.
+    globalNames.resize(strings.size());
+
+    // Find globals to reuse (see comment on stringPtrsToPreserve for context).
+    for (auto& global : module->globals) {
+      if (global->type == nnstringref && !global->imported()) {
+        if (auto* stringConst = global->init->dynCast<StringConst>()) {
+          auto stringIndex = stringIndexes[stringConst->string];
+          auto& globalName = globalNames[stringIndex];
+          if (!globalName.is()) {
+            // This is the first global for this string, use it.
+            globalName = global->name;
+            stringPtrsToPreserve.insert(&global->init);
+          }
+        }
+      }
+    }
+
+    Builder builder(*module);
+    for (Index i = 0; i < strings.size(); i++) {
+      auto& globalName = globalNames[i];
+      if (globalName.is()) {
+        // We are reusing a global for this one.
+        continue;
+      }
+
+      auto& string = strings[i];
+      auto name = Names::getValidGlobalName(
+        *module, std::string("string.const_") + std::string(string.str));
+      globalName = name;
+      newNames.insert(name);
+      auto* stringConst = builder.makeStringConst(string);
+      auto global =
+        builder.makeGlobal(name, nnstringref, stringConst, Builder::Immutable);
+      module->addGlobal(std::move(global));
+    }
+
+    // Sort our new globals to the start, as others may use them.
+    std::stable_sort(
```

Comment:
Existing globals may also have many uses as well. Is there a principled reason to expect the new globals to be used more frequently?

Reply:
This is for correctness, not compactness. This sort just makes sure, in the simplest way I can think of, that our new globals are in a valid position. We leave frequency sorting for `reorder-globals` (which we run after this).

Reply:
(And to answer the question, in case I've misunderstood your intent: no, there's no reason to expect the new globals to be more or less used than the old. It's just that the old are already in a proper position.)

Reply:
Oh, I see. Can you expand the comment to make this more explicit? Perhaps "Sort our new globals to the start to ensure validity, as existing globals initializers may reference them."